modules/boulder/src/main/ts/ephox/boulder/core/ValuePresence.ts

```diff
@@ -0,0 +1,52 @@
+import * as FieldPresence from '../api/FieldPresence';
+import { Processor } from './ValueProcessor';
+
+interface FieldData {
+  readonly key: string;
+  readonly newKey: string;
+  readonly presence: FieldPresence.FieldPresenceTypes;
+  readonly prop: Processor;
+}
+
+interface StateData {
+  readonly newKey: string;
+  readonly instantiator: (obj: any) => unknown;
+}
+
+interface ValuePresenceData<D, T> {
+  readonly discriminator: D;
+  readonly data: T;
+}
+
+export type FieldProcesserData = ValuePresenceData<'field', FieldData>;
+export type StateProcessorData = ValuePresenceData<'state', StateData>;
+
+export type ValueProcessorTypes = FieldProcesserData | StateProcessorData;
+
+// These are the types for the fold callbacks
+export type FieldValueProcessor<T> = (key: string, newKey: string, presence: FieldPresence.FieldPresenceTypes, prop: Processor) => T;
+export type StateValueProcessor<T> = (newKey: string, instantiator: (obj: any) => any) => T;
+
+const constructors = {
```

Comment:
Why make this object then unpack it and export it. Do you even need this object? I'm guessing it's there to try to mimic adt constructors but simple make functions would be the same thing.

Reply:
Good point. I copied another ADT conversion that did need it, but we really don't here. Will change it.